### PR TITLE
Make git material checking out working copy only when necessary

### PR DESCRIPTION
**Pull request by wpc**
For #1890. This PR helps saving disk space on server and speedup checking out both on server and agent.

### CODE REVIEW

**Comment by arvindsv** [comment]
Looks ok to me. This should affect only new clones. Is there a verification missing in the tests?

File: common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java [hunk 1]
```diff
@@ -59,7 +59,11 @@ public int cloneFrom(ProcessOutputStreamConsumer outputStreamConsumer, String ur
     // Clone repository from url with specified depth.
     // Special depth 2147483647 (Integer.MAX_VALUE) are treated as full clone
     public int cloneFrom(ProcessOutputStreamConsumer outputStreamConsumer, String url, Integer depth) {
-        CommandLine gitClone = git().withArg("clone").withArg(String.format("--branch=%s", branch));
+        CommandLine gitClone = git()
```
Comment (ketan) [inline common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: This feels a bit odd in that it claims to say clone, and does not behave like `git clone` can we make the `noCheckout` a boolean arg?

Also, don't we need a clone with checkout on the agent?

Comment (arvindsv) [inline common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: Yeah, you're right (about the agent).

Comment (wpc) [inline common/src/com/thoughtworks/go/domain/materials/git/GitCommand.java]: @ketan 
On agent we always do a hard reset to a specific revision, so the initial working copy checkout is not needed and just wasting time.

**Pull request merged**